### ISSUE TITLE
Improve spatial_mean

### ISSUE THREAD

**Issue by aulemahal** [Ouranosinc/xscen#94]
### Generic Issue

While it works well, there are some caveats.

- [x] "Global" mean with xESMF : We need a way to specify we want to average over the whole grid with the xESMF method.
- [x] "cos(lat)" weighting : For the non-xESMF methods, it could be interesting to be able to weights the grid cells according to their real size with the easy `cos(lat)` approximation.
- [x] xESMF inputs : We need to re-implement the possibility to pass a `GeoDataframe` as a "shape".
- [ ] xESMF : Better error message when polygon cleaning is needed., i.e. : when `gdf.simplify(tolerance=XX, preserve_topology=True).buffer(0)` is needed. Could also be done automatically, but we'd need a way to guess the tolerance (it depends on the polygon's CRS, on the grid's resolution and on the target precision).
- [x] Segmentize polygons if needed before sending to xESMF and remove warning.
- [x] Expose xESMF's `out_chunks` arg for the regridder (xESMF 0.8)

**Comment by RondeauG** [comment]
Currently, a workaround would be to use smaller polygons. For example, this uses 1 rectangle per 1° of longitude and gives the right answer:
```
eps = 1e-4
p = []
for lon in range(-180, 180):
    p.append(Polygon([(lon, -90 + eps), (lon, 90- eps), (lon + 1, 90- eps), (lon + 1, -90 + eps)]))

savg = xe.SpatialAverager(ds, p)
savg.weights.max()

savg(ds.tg_mean).mean('geom') - 273.15
```

Another solution is to circumvent `SpatialAverager` completely and use ESMF (not `xESMF`) to find the area covered by each pixel:
```
from ESMF import Field
import xesmf as xe

grid_in, _, crds = xe.frontend.ds_to_ESMFgrid(ds, need_bounds=True)
area_raw = Field(grid_in, name='area')
area_raw.get_area()

area = xr.DataArray(area_raw.data, dims=crds[::-1], coords={c: ds[c] for c in crds})

ds.tg_mean.weighted(dsa.area).mean(['lat', 'lon']) - 273.15
```

**Comment by huard** [comment]
If you find bugs with xESMF, please open [issues](https://github.com/pangeo-data/xESMF/issues).

**Comment by RondeauG** [comment]
Indeed, I've been talking to Pascal regarding this. We ideally want to fix the issue in xESMF, I just haven't had the time to make the Issue yet.

**Comment by aulemahal** [comment]
The xESMF issue has been "fixed" in 0.8. In fact, the fix is simply to use a function from shapely 2.0, so we might be able to close this issue without even pinning xESMF.

**Comment by RondeauG** [comment]
Go ahead with what you think is best, you're the more up-to-date person on that issue!
